Document custom derive.

steveklabnik opened the pull request:
These are some bare-bones documentation for custom derive, needed
to stabilize "macros 1.1",
https://github.com/rust-lang/rust/issues/35900

The book chapter is based off of a blog post by @cbreeden,
https://cbreeden.github.io/Macros11/

Normally, we have a policy of not mentioning external crates in
documentation. However, given that syn/quote are basically neccesary
for properly using macros 1.1, I feel that not including them here
would make the documentation very bad. So the rules should be bent
in this instance.

So far, this PR includes only docs; @alexcrichton said in https://github.com/rust-lang/rust/issues/35900 that he'd be okay with landing them before stabilization; I don't mind either way.

steveklabnik commented:
/cc @rust-lang/docs @rust-lang/lang

est31 commented:
Would it make sense to document it in the reference as well?

steveklabnik commented:
> Would it make sense to document it in the reference as well? Derive is documented there too.

Ah ha! I knew I forgot something. Yes.

killercup commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,194 @@
+With some kind of nice output, like `Hello, World! My name is Pancakes.`. 
+
+Let's go ahead and write up what we think our macro will look like from a user
+perspective. In `hello-world\main.rs` we write:
```
Shouldn't that be `src/main.rs`?

killercup commented on src/doc/book/procedural-macros.md:
````diff
@@ -0,0 +1,194 @@
+```
+
+Great. So now we just need to actually write the procedural macro. Let's start
+a new crate called `hello-world-macro` inside our `hello-world` project.
````
Maybe add a sentence about naming convention? I've seen a bunch of crates that call these `<name>_derive`.

killercup commented on src/doc/book/procedural-macros.md:
````diff
@@ -0,0 +1,194 @@
+```
+
+So there is a lot going on here.  We have introduced two new crates: `syn` and
+`quote`.  As you may have noticed, `input: TokenSteam` is immediately converted
````
Link "syn" and "quote" to their repositories or crates.io pages?

killercup commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,194 @@
+to a String.  This String is a string representation of the rust code for which
+we are deriving `HelloWorld` for--that's it.  So what we really need is to be
+able to _parse_ rust code into something usable.  This is where `syn` comes to
+play.  `syn` is a `nom` based rust parser, and it's extremely useful.  The
```
People who don't know nom will be confused by this.

killercup commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,194 @@
+
+So there is a lot going on here.  We have introduced two new crates: `syn` and
+`quote`.  As you may have noticed, `input: TokenSteam` is immediately converted
+to a String.  This String is a string representation of the rust code for which
```
Lowercase "rust", same a few times below

killercup replied:
Put "String" in back ticks?

killercup commented on src/doc/book/procedural-macros.md:
````diff
@@ -0,0 +1,194 @@
+}
+```
+
+So there is a lot going on here.  We have introduced two new crates: `syn` and
````
Double space after sentence here and a few times below; I don't think the rest of the book does this.

killercup commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,194 @@
+able to _parse_ rust code into something usable.  This is where `syn` comes to
+play.  `syn` is a `nom` based rust parser, and it's extremely useful.  The
+other crate we've introduced is `quote`.  It's essentially the dual of `syn` as
+it will make serializing rust code really easy.
```
Hm, maybe "serializing rust code" → "generating Rust code"?

killercup commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,194 @@
+    let s = input.to_string();
+    
+    // Parse the string representation
+    let ast = syn::parse_macro_input(&s).unwrap();
```
Error handling probably deserves a short paragraph as well, even if it's just "panicking the proc macro code will show it as a compiler error point to the derive".

withoutboats commented:
How do you feel about making `HelloWorld` into a trait? Implementing traits is intended as the normal use case for derive, and it seems odd to document it with a non-trait impl (though of course that does work).

steveklabnik commented:
Okay! I think I've fixed up everything, @withoutboats @killercup . Thanks for the review 😄 

Still have to update the reference, should push that in a minute.

killercup commented on src/doc/book/procedural-macros.md:
````diff
@@ -146,45 +161,45 @@ fn impl_hello_world(ast: &syn::MacroInput) -> quote::Tokens {
 ```bash
 error: the `#[proc_macro_derive]` attribute is only usable with crates of the `proc-macro` crate type
- --> hello-world-macro\src\lib.rs:8:3
+ --> hello-world-derive\src\lib.rs:8:3
````
FYI this path uses backslashes, the rest of the book probably not

frewsxcv commented on src/doc/reference.md:
```diff
@@ -555,26 +555,22 @@ mod a {
   declarative way.
+* [Procedural Macros][procedural macros] can be used to implement custom derive.
 
 ## Macros
```
It's a little weird to me that `## Macros` is a subheading within `# Macros`, but I see why you did it

steveklabnik replied:
yeah i would like to re-do this and other parts of the reference but i'm trying to be minimally invasive here

frewsxcv commented on src/doc/reference.md:
````diff
@@ -2319,6 +2337,8 @@ impl<T: PartialEq> PartialEq for Foo<T> {
 }
 ```
 
+You can implement derive for your own type through [procedural macros](#procedural-macros).
````
Earlier in this section, "derive" is styled as

```
`derive`
```

with backticks, so might be good to be consistent.

frewsxcv commented on src/doc/reference.md:
```diff
@@ -4122,6 +4142,16 @@ be ignored in favor of only building the artifacts specified by command line.
+* `--crate-type=proc-macro`, `#[crate_type = "proc-macro"]` - The output
+  produced is not specified, but if a `-L` path is provided to it then the
+  compiler will recognize the output artifacts as a macro and it can be loaded
+  for a program. If a crate is compiled with the proc-macro crate type t will
```
```
s/t will/it will/
```

It looks like crate types are styled with backticks, so also:

```
s/proc-macro/`proc-macro`/
```

frewsxcv commented on src/doc/reference.md:
```diff
@@ -4122,6 +4142,16 @@ be ignored in favor of only building the artifacts specified by command line.
+  forbid exporting any items in the crate other than those functions tagged
+  `#[proc_macro_derive]` and those functions must also be placed at the crate
+  root. Finally, the compiler will automatically set the `cfg(proc_macro)`
+  annotation whenever any crate type of a compilation is the rustc-macro crate
```
```
s/rustc-macro/`rustc-macro`/
```

nikomatsakis commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,209 @@
+
+Rust includes several traits that you can derive, but it also lets you define
+your own. We can accomplish this task through a feature of Rust called
+"procedrual macros." Eventually, procedural macros will allow for all sorts of
```
typo: procedural

nikomatsakis commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,209 @@
+type using `ast.ident`. The `quote!` macro let's us write up the Rust code
+that we wish to return and convert it into `Tokens`. `quote!` let's us use some
+really cool templating mechanics; we simply write `#name` and `quote!` will
+replace it with the variable named `name`. You can even do some repition
```
Nit: s/repition/repetition/

nikomatsakis commented on src/doc/reference.md:
```diff
@@ -652,6 +648,28 @@ Rust syntax is restricted in two ways:
 
 [RFC 550]: https://github.com/rust-lang/rfcs/blob/master/text/0550-macro-future-proofing.md
 
+## Procedrual Macros
```
Nit: s/Procedrual/Procedural/

nikomatsakis commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,209 @@
+% Procedrual Macros (and custom Derive)
```
Nit: s/Procedrual/Procedural/

steveklabnik replied:
i am... really bad at this spelling apparently, haha

nikomatsakis commented on src/doc/book/procedural-macros.md:
````diff
@@ -0,0 +1,209 @@
+```
+
+Oh, so it appears that we need to declare that our `hello-world-derive` crate is
+a `proc-macro` crate type. How do we do this? Like this:
````
This stream of consciousness bit felt a *little* over the top to me, but it's fine. =)

steveklabnik commented:
> The one thing I would say is that I would have expected a bit more up-front material discussing the 'basic model' for derive

I'm not totally opposed to this; I was thinking about making it future proof. That is, right now it's string in string out because it's a hack. It's really `TokenStream -> String -> process -> String -> TokenStream`, and ideally those string bits are going away. But I guess it will be a long time...

chris-morgan commented on src/doc/book/SUMMARY.md:
```diff
@@ -52,6 +52,7 @@
     * [Borrow and AsRef](borrow-and-asref.md)
     * [Release Channels](release-channels.md)
     * [Using Rust without the standard library](using-rust-without-the-standard-library.md)
+    * [Procedural Macros (and custom Derive)](procedural-macros.md)
```
I see we’re continuing the tradition of inconsistent capitalisation—nay, taking it to a new level, with mixed title and sentence case all in one line!

steveklabnik replied:
Yeah "Derive" felt like something that deserved a capital here, maybe not, idk

F001 commented:
Thanks for writing the doc. I can get the basic concept of proc-macro now, but still haunted by some questions.

1. Is it possible to write this macro without any third party dependency? What if we create it from scratch? What kind of benefit we can get from syn&quote crate?
2. Why does `TokenStream` type exist, if the only purpose of it is to be converted to a `String`? It is weird to convert a structured type to a `String`, and then parse. I would like to see more explanation.

withoutboats commented on src/doc/reference.md:
```diff
@@ -555,26 +555,22 @@ mod a {
 
-Users of `rustc` can define new syntax extensions in two ways:
-
-* [Compiler plugins][plugin] can include arbitrary Rust code that
```
We should probably continue to document compiler plugins in the reference for the time being; they still exist & will for some time, and of course some project are heavily reliant on them.

steveklabnik replied:
This isn't removing documentation; it's re-ordering it. I thought putting macros by example first was a better way of doing things.

withoutboats replied:
It looks like it went from `Compiler plugins, Macros` to `Macros, Procedural Macros`, shouldn't it be `Macros, Procedural Macros, Compiler plugins`?

steveklabnik replied:
My understanding is that procedural macros are compiler plugins.

withoutboats replied:
That's correct, but we've implemented (and now stabilized) the new procedural macros interface only for controlling derive macros. Users who want to define custom attributes or function style procedural macros (without using a hack) are still using the never-to-be-stabilized compiler plugin interface.

Until we've fully implemented procedural macros, compiler plugins will still be used by these users and so we shouldn't drop whatever docs we have on it (even if we don't focus any attention on improving it).

steveklabnik replied:
I didn't actually remove any of those docs generally, just the small description here. Let me fix it.

parir commented on src/doc/book/procedural-macros.md:
````diff
@@ -0,0 +1,209 @@
+```
+
+To make sure that our `hello-world` crate is able to find this new crate we've
+crated, we'll add it to our toml:
````
s/crated/created/

steveklabnik commented:
@F001 

> Is it possible to write this macro without any third party dependency? What if we create it from scratch? What kind of benefit we can get from syn&quote crate?

Sure, but it's much, much, much harder. The benefits are that syn and quote do parsing and generation for you.

> Why does TokenStream type exist, if the only purpose of it is to be converted to a String? It is weird to convert a structured type to a String, and then parse.

This is the difference between macros 1.1 and macros 2.0, in a nutshell. 

>  I would like to see more explanation about what kind of API does the proc-macro feature provide for us, not the API of a third party lib.

This is the entire API. That's it. There's nothing else to explain. It is extremely bare-bones.

steveklabnik commented:
Okay, I believe that I've addressed everyone's comments now, both on GitHub and in the text. I've squashed, since this is going to need to be backported, and want to make that easy.

nikomatsakis commented:
Let's do it.

@bors r+

bors commented:
:pushpin: Commit eb2c83f has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis eb2c83f7026d40966bfa1314ee875bc81196c1e2 -->

nikomatsakis commented:
@bors rollup

steveklabnik commented:
@bors: r=nikomatsakis rollup  (https://github.com/rust-lang/rust/pull/38770#issuecomment-270183013)

bors commented:
:pushpin: Commit c0efdbf has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis c0efdbf8e5c6c5805d0f2e603690a5ce0f706b92 -->

alexcrichton commented:
@bors: rollup- p=1

Let's land this quickly so we can backport

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #38783) made this pull request unmergeable. Please resolve the merge conflicts.

steveklabnik commented:
@bors: r=alexcrichton p=1

bors commented:
:pushpin: Commit 3075c1f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 3075c1f65e08e0b52dcf872588358daffef8b47c -->

sgrif commented on src/doc/book/procedural-macros.md:
```diff
@@ -0,0 +1,213 @@
+moment, procedural macros need to be in their own crate. Eventually, this
+restriction may be lifted, but for now, it's required. As such, there's a
+convention; for a crate named `foo`, a custom derive procedural macro is called
+`foo-derive`. Let's start a new crate called `hello-world-derive` inside our
```
I think the convention is `foo_derive` not `foo-derive`.

steveklabnik replied:
package names are supposed to prefer `-` over `_`.

bors commented:
:hourglass: Testing commit 3075c1f with merge 8b9bc29...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/189016572)

steveklabnik commented:
@bors: retry

(travis is spurious)

steveklabnik commented:
@bors: force

bors commented:
:hourglass: Testing commit 3075c1f with merge 5d994d8...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1396), [status-travis](https://travis-ci.org/rust-lang/rust/builds/189031533)
Approved by: alexcrichton
Pushing 5d994d8b7e482e87467d4a521911477bd8284ce3 to master...
